contracts/token/ERC20Rebasable.sol

```diff
@@ -0,0 +1,377 @@
+    /// @param spender_ An address of the spender of the tokens
+    /// @param amount_ An amount of allowance spend
+    function _spendAllowance(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal {
+        uint256 currentAllowance = _getTokenAllowance()[owner_][spender_];
+        if (currentAllowance == type(uint256).max) {
+            return;
+        }
+        if (amount_ > currentAllowance) {
+            revert ErrorNotEnoughAllowance();
+        }
+        unchecked {
+            _approve(owner_, spender_, currentAllowance - amount_);
+        }
+    }
+
+    /// @dev Sets amount_ as the allowance of spender_ over the owner_'s tokens
+    /// @param owner_ An address of the account to set allowance
+    /// @param spender_ An address of the tokens spender
+    /// @param amount_ An amount of tokens to allow to spend
+    function _approve(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal virtual onlyNonZeroAccount(owner_) onlyNonZeroAccount(spender_) {
+        _getTokenAllowance()[owner_][spender_] = amount_;
+        emit Approval(owner_, spender_, amount_);
+    }
+
+    function _sharesOf(address account_) internal view returns (uint256) {
+        return _getShares()[account_];
+    }
+
+    function _getTokensByShares(uint256 sharesAmount_) internal view returns (uint256) {
+        (uint256 tokensRate, uint256 decimals)  = _getTokensRateAndDecimal();
+        return (sharesAmount_ * tokensRate) / (10 ** decimals);
+    }
+
+    function _getSharesByTokens(uint256 tokenAmount_) internal view returns (uint256) {
+        (uint256 tokensRate, uint256 decimals)  = _getTokensRateAndDecimal();
+        return (tokenAmount_ * (10 ** decimals)) / tokensRate;
+    }
+
+    function _getTokensRateAndDecimal() internal view returns (uint256, uint256) {
+        uint8 rateDecimals = TOKEN_RATE_ORACLE.decimals();
+
+        if (rateDecimals == uint8(0) || rateDecimals > uint8(18)) revert ErrorInvalidRateDecimals(rateDecimals);
+
+        //slither-disable-next-line unused-return
+        (,
+        int256 answer
+        ,
+        ,
+        uint256 updatedAt
+        ,) = TOKEN_RATE_ORACLE.latestRoundData();
+
+        if (updatedAt == 0) revert ErrorWrongOracleUpdateTime();
+        if (answer <= 0) revert ErrorOracleAnswerIsNegative();
+
+        return (uint256(answer), uint256(rateDecimals));
+    }
+
+    /// @dev Creates amount_ shares and assigns them to account_, increasing the total shares supply
+    /// @param recipient_ An address of the account to mint shares
+    /// @param amount_ An amount of shares to mint
+    function _mintShares(
+        address recipient_,
+        uint256 amount_
+    ) internal onlyNonZeroAccount(recipient_) returns (uint256)  {
+        _setTotalShares(_getTotalShares() + amount_);
+        _getShares()[recipient_] = _getShares()[recipient_] + amount_;
+        emit Transfer(address(0), recipient_, amount_);
```

Comment:
See the events in L1 contracts:
```solidity
    function _emitTransferEvents(address _from, address _to, uint _tokenAmount, uint256 _sharesAmount) internal {
        emit Transfer(_from, _to, _tokenAmount);
        emit TransferShares(_from, _to, _sharesAmount);
    }
```

There must be two alike events here as well: one about wstETH token transfer, and one about StETH token transfer (minting) with StETH not shares amount.